### PR TITLE
Depth First Search Python

### PR DESCRIPTION
**Pull request by soumyavemuri**
### Have you read the [Contributing Guidelines on Pull Requests](https://github.com/TesseractCoding/NeoAlgo/blob/master/CONTRIBUTING.md#pull-requests)?

Yes.

### Description

Depth-First Search Traversal in Python was corrected and updated.

### Checklist

- [x] I've read the contribution guidelines.
- [x] I've checked the issue list before deciding what to submit.
- [x] I've edited the `README.md` and link to my code.

## Related Issues or Pull Requests

Issue #2398

### CODE REVIEW

**Comment by soumyavemuri** [comment]
@bhav09 @ankitaggarwal23 @gat786 Please review when possible, and let me know if any changes are required.

File: Python/search/dfs.py [hunk 1]
```diff
@@ -1,58 +1,111 @@
...
+ret = G.dfs_helper(start_node)
+
+# If the user entered a non-existent vertiex as start node, the following is implemented
+# 1 is returned if the vertex is non existent
+while ret==1:
+    start_node = input('\n'+"Enter the Start Node for traversal: ")
+    ret = G.dfs_helper(start_node)
+
+# To implement dfs for bi-directional graphs, replace G.add_edge in lines 88-96
+# with G.add_edge_bidirectional
```
Comment (gat786) [inline Python/search/dfs.py]: Always have a empty line at the end of your file.

File: Python/search/dfs.py [hunk 2]
```diff
@@ -1,58 +1,111 @@
...
+            
+    def dfs(self,node):
+        # dfs function
+        if node not in self.visited:            # checks if the node was already visited
+            self.traversal.append(node)         # if the node wasnt traversed already, it's appended to the traversal
+            self.visited.add(node)              # the node is marked visited
+            for neighbour in self.graph[node]:  # the neighbouring vertices of the node are explored one after the other     
+                self.dfs(neighbour)
+        
+# Code Implementation starts here
```
Comment (gat786) [inline Python/search/dfs.py:84]: keep all the controller code inside 

```py
if __name___ == "__main__":
```
code block

File: Python/search/dfs.py [hunk 3]
```diff
@@ -1,58 +1,111 @@
...
+
+# Graph instance is created
+G = Graph()
+
+# vertices are added to the graph
+for v in range(int(input('How many vertices are in the graph? '))):
```
Comment (gat786) [inline Python/search/dfs.py]: you can simplify this line of code by taking input on another line and then putting it inside of loop. 
Readability of Code is more important than SLOC count.

Comment (gat786) [reply Python/search/dfs.py]: Same goes for multiple lines below this

File: Python/search/dfs.py [hunk 4]
```diff
@@ -1,58 +1,111 @@
...
+        source,destination = input("Enter Source and Destination Vertices separated by a space: ").split()
+        val = G.add_edge(source,destination)
+
+# start node for the dfs traversal is taken from the user
+start_node = input('\n'+"Enter the Start Node for traversal: ")
+
+# dfs is implemented
+ret = G.dfs_helper(start_node)
+
+# If the user entered a non-existent vertiex as start node, the following is implemented
```
Comment (gat786) [inline Python/search/dfs.py]: `vertices` and not `vertiex`

File: Python/search/dfs.py (hunk 1 above)
Comment (gat786) [reply Python/search/dfs.py]: Add an example of how someone would run this code at the very end of this code file. 
Also add space and time complexities for the algorithm you have implemented.

**Review by gat786 — CHANGES_REQUESTED** [review]

**Comment by soumyavemuri** [comment]
@gat786 Thank you very much for the useful feedback. Thank your taking the time to give a detailed review.

I have made the requested changes.

File: Python/search/dfs.py [hunk 5]
```diff
@@ -1,58 +1,165 @@
...
+    def add_vertex(self, vertex):
+        # vertex is added to the graph dictionary and it's adjacency list is declared
+        self.graph[vertex] = []
+    
+    def add_edge(self, source, destination):
+        # directional edges are added in this function
+        # the edge is directional, i.e., going from source to destination
+
+        # conditional statements in case the vertex aren't present in the graph
+        if destination not in self.graph or source not in self.graph:
```
Comment (gat786) [inline Python/search/dfs.py:21]: This condition check becomes useless when you check the samethings again inside the code block.

File: Python/search/dfs.py [hunk 6]
```diff
@@ -1,58 +1,165 @@
...
+
+        # edge is added to the graph, from source to destination
+        self.graph[source].append(destination)
+
+    def add_edge_bidirected(self, vertex1, vertex2):
+        # bidirectional graph edges are added in this function
+        # the edge is bidirectional, i.e., the edge goes both ways
+
+        # conditional statements to check if the vertices are present
+        if vertex2 not in self.graph or vertex1 not in self.graph:
```
Comment (gat786) [inline Python/search/dfs.py:39]: This condition is also useless if you check it again inside the block one by one. 

You can remove this parent if statement and your code will perform in same way

File: Python/search/dfs.py [hunk 7]
```diff
@@ -1,58 +1,165 @@
...
+    def dfs_helper(self, start):
+        # function to assist in calling dfs
+        # also prints the final dfs traversal
+
+        # checks if the start node is in the graph
+        if start not in self.graph:
+            print('\n'+str(start)+' not in graph.'+'\n')
+            # 1 is returned to avoid stopping the running of the program,
+            # allows the user to re-enter the start vertex
+            return 1
```
Comment (gat786) [inline Python/search/dfs.py:60]: Instead of returning 1 when there is an exception you can throw exceptions

Comment (gat786) [reply Python/search/dfs.py:60]: Also, it looks you are putting in a lot of efforts and it's always good to see such enthusiastic contributors and work with them.

**Review by gat786 — CHANGES_REQUESTED** [review]

**Review by gat786 — APPROVED** [review]

**Review by ankitaggarwal23 — APPROVED** [review]

File: Python/search/dfs.py (hunk 7 above)
Comment (soumyavemuri) [reply Python/search/dfs.py:60]: Thank you for your guidance. I have learnt quite a lot with just this one PR. Thank you so much for that.

**Comment by soumyavemuri** [comment]
@HarshCasper When you get time, can you add the level tag, please?

**Pull request merged**